### PR TITLE
ARROW-13965: [C++] dynamic_casts in parquet TypedColumnWriterImpl impacting performance

### PR DESCRIPTION
**Pull request by etseidl**
This PR adds observer pointers to TypedColumnWriterImpl to avoid repeated dynamic_casts upon writing many small batches.  This can lead to significant performance gains.  As an example, using stream_reader_writer from examples/parquet/parquet_stream_api modified to write 20M rows, an improvement of over 30% was seen in write times.
```
original:
52.919u 2.146s 0:55.81 98.6%	0+0k 0+2646928io 0pf+0w
53.328u 2.327s 0:56.38 98.6%	0+0k 0+2646928io 0pf+0w
53.048u 2.341s 0:56.06 98.7%	0+0k 0+2646928io 0pf+0w

patched:
35.367u 2.012s 0:37.88 98.6%	0+0k 0+2646928io 0pf+0w
35.133u 1.717s 0:37.36 98.6%	0+0k 0+2646928io 0pf+0w
35.228u 1.950s 0:37.61 98.8%	0+0k 0+2646928io 0pf+0w
```

### CODE REVIEW

File: cpp/src/parquet/column_writer.cc [hunk 1]
```diff
@@ -1358,6 +1363,8 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
       // Only PLAIN encoding is supported for fallback in V1
       current_encoder_ = MakeEncoder(DType::type_num, Encoding::PLAIN, false, descr_,
                                      properties_->memory_pool());
+      current_value_encoder_ = dynamic_cast<ValueEncoderType*>(current_encoder_.get());
+      current_dict_encoder_ = 0;  // not using dict
```
Comment (emkornfield) [inline cpp/src/parquet/column_writer.cc:1367]:
```suggestion
      current_dict_encoder_ = nullptr;  // not using dict
```

**Review by emkornfield — COMMENTED** [review]
Is it possible to add a benchmark to show the improvement?  Or is there an existing one that shows this?

**Comment by pitrou** [comment]
TBH, the stream APi is probably not high-performance at all. I'm not sure it makes sense to write benchmarks for it (unless we want to showcase how slow it is :-)).

**Comment by etseidl** [comment]
Right now I'm trying to modify the stream_reader_writer example to see if the speed bump is noticeable there.  If not, I'll  modify my code (which doesn't actually use the streaming api, it just writes in batches of 1) to  do writes to a fixed schema.

**Comment by etseidl** [comment]
Ok, I modified the stream_reader_writer example to just do the write, and changed the number of rows from 2000 to 20M (and the max row group size from 1000 to 10M).  I ran this on my laptop with and without the changes in this patch.  I used the csh time command to compare run times. The results are:
```
original:
52.919u 2.146s 0:55.81 98.6%	0+0k 0+2646928io 0pf+0w
53.328u 2.327s 0:56.38 98.6%	0+0k 0+2646928io 0pf+0w
53.048u 2.341s 0:56.06 98.7%	0+0k 0+2646928io 0pf+0w

patched:
35.367u 2.012s 0:37.88 98.6%	0+0k 0+2646928io 0pf+0w
35.133u 1.717s 0:37.36 98.6%	0+0k 0+2646928io 0pf+0w
35.228u 1.950s 0:37.61 98.8%	0+0k 0+2646928io 0pf+0w
```
@emkornfield Do you want benchmark code added to the PR, or are these results sufficient?

**Comment by pitrou** [comment]
As far as am I concerned (and not wanting to speak for @emkornfield  :-)), I would be ok with simply adding these numbers to the PR description.

**Comment by etseidl** [comment]
Sorry, I kept the PR description short since I had more details in the JIRA issue.  Once @emkornfield concurs, I'll edit the description.  Thanks!

**Comment by emkornfield** [comment]
Yes, I think just updating the description is fine.

**Review by pitrou — APPROVED** [review]
Thank you @etseidl . I'll merge this if CI passes.

**Comment by pitrou** [comment]
Note the LZ4 build failures on MinGW are unrelated.

**Pull request closed without merging**